# How to Filter Pandas DataFrames with `query()`:

This guide shows multiple examples of how to use the Pandas `query()` function to filter meteorological data.  The `query()` function provides a clean, readable way to filter DataFrames using string expressions.

## Sample Dataset Setup

First, let's create a sample weather dataset:

In [1]:
import pandas as pd
import numpy as np

# Set random seed for reproducibility
np.random.seed(42)

# Create sample meteorological data
data = {
    'date': pd.date_range('2025-01-01', periods=100, freq='H'),
    'temperature': np.random.uniform(-10, 35, 100),  # °C
    'humidity': np.random.uniform(20, 100, 100),     # %
    'pressure': np.random. uniform(980, 1030, 100),   # hPa
    'wind_speed': np.random.uniform(0, 25, 100),     # m/s
    'precipitation': np.random.uniform(0, 10, 100),  # mm
    'station':  np.random.choice(['KMSP', 'KORD', 'KJFK', 'KSEA'], 100)
}

df = pd.DataFrame(data)

# Display the first few rows
print("Dataset shape:", df.shape)
df.head(10)

Dataset shape: (100, 7)


/var/folders/51/b9llyv5s4dd7llx0xjt1rzkw0000gr/T/ipykernel_47414/1983085336.py:9: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  'date': pd.date_range('2025-01-01', periods=100, freq='H'),


,date,temperature,humidity,pressure,wind_speed,precipitation,station
0,2025-01-01 00:00:00,6.854305,22.514335,1012.101582,1.292043,1.031239,KORD
1,2025-01-01 01:00:00,32.782144,70.912833,984.206998,13.283866,9.025529,KJFK
2,2025-01-01 02:00:00,22.939727,45.148478,988.081436,13.515878,5.052524,KMSP
3,2025-01-01 03:00:00,16.939632,60.685655,1024.927709,15.935748,8.264575,KMSP
4,2025-01-01 04:00:00,-2.979161,92.605318,1010.321453,18.152283,3.200496,KMSP
5,2025-01-01 05:00:00,-2.980247,39.943378,980.459853,24.396302,8.955232,KMSP
6,2025-01-01 06:00:00,-7.386237,52.830634,985.073577,12.907509,3.892017,KSEA
7,2025-01-01 07:00:00,28.977927,80.444091,1013.175088,8.073912,0.108377,KMSP
8,2025-01-01 08:00:00,17.050176,38.303853,980.253079,19.879655,9.053820,KSEA
9,2025-01-01 09:00:00,21.863266,26.158393,988.040403,6.770806,0.912867,KSEA


## Basic Filtering Examples

### 1. Simple Comparison
Filter for temperatures above freezing:

In [2]:
# Find all records where temperature is above 0°C
warm_weather = df.query('temperature > 0')

print(f"Total records: {len(df)}")
print(f"Warm weather records: {len(warm_weather)}")
warm_weather.head()

Total records: 100
Warm weather records: 71


,date,temperature,humidity,pressure,wind_speed,precipitation,station
0,2025-01-01 00:00:00,6.854305,22.514335,1012.101582,1.292043,1.031239,KORD
1,2025-01-01 01:00:00,32.782144,70.912833,984.206998,13.283866,9.025529,KJFK
2,2025-01-01 02:00:00,22.939727,45.148478,988.081436,13.515878,5.052524,KMSP
3,2025-01-01 03:00:00,16.939632,60.685655,1024.927709,15.935748,8.264575,KMSP
7,2025-01-01 07:00:00,28.977927,80.444091,1013.175088,8.073912,0.108377,KMSP


### 2. Multiple Conditions (AND)
Filter for ideal flying conditions:

In [3]:
# Low wind and good visibility conditions
good_flying = df.query('wind_speed < 10 and humidity < 70')

print(f"Good flying conditions: {len(good_flying)} records")
good_flying.head()

Good flying conditions: 27 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station
0,2025-01-01 00:00:00,6.854305,22.514335,1012.101582,1.292043,1.031239,KORD
9,2025-01-01 09:00:00,21.863266,26.158393,988.040403,6.770806,0.912867,KSEA
11,2025-01-01 11:00:00,33.645943,32.897703,1014.594760,1.961410,9.500620,KMSP
17,2025-01-01 17:00:00,13.614039,34.925605,1017.324570,4.332358,3.286645,KJFK
19,2025-01-01 19:00:00,3.105311,63.147379,1022.461171,6.256072,7.523745,KMSP


### 3. Multiple Conditions (OR)
Filter for extreme conditions:

In [4]:
# Very hot or very cold conditions
extreme_temps = df.query('temperature > 30 or temperature < -5')

print(f"Extreme temperature records: {len(extreme_temps)}")
extreme_temps.head()

Extreme temperature records: 24


,date,temperature,humidity,pressure,wind_speed,precipitation,station
1,2025-01-01 01:00:00,32.782144,70.912833,984.206998,13.283866,9.025529,KJFK
6,2025-01-01 06:00:00,-7.386237,52.830634,985.073577,12.907509,3.892017,KSEA
10,2025-01-01 10:00:00,-9.073698,43.180116,1007.436689,10.974286,3.193136,KMSP
11,2025-01-01 11:00:00,33.645943,32.897703,1014.594760,1.961410,9.500620,KMSP
29,2025-01-02 05:00:00,-7.909731,60.859784,1011.556931,6.193275,1.170670,KMSP


### 4. Range Filtering
Filter for comfortable weather conditions:

In [5]:
# Temperature between 18-25°C and low wind
comfortable = df.query('18 <= temperature <= 25 and wind_speed < 5')

print(f"Comfortable conditions: {len(comfortable)} records")
comfortable.head()

Comfortable conditions: 3 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station
70,2025-01-03 22:00:00,24.751015,74.205149,1020.468058,2.519450,0.121545,KMSP
74,2025-01-04 02:00:00,21.808580,71.613823,1005.567120,1.779716,5.277011,KMSP
87,2025-01-04 15:00:00,18.690086,64.416065,1007.132232,4.435989,2.809632,KSEA


## Intermediate Examples

### 5. String Matching
Filter by weather station:

In [6]:
# Get data from Chicago O'Hare
chicago_data = df.query("station == 'KORD'")

print(f"Chicago O'Hare records: {len(chicago_data)}")
chicago_data.head()

Chicago O'Hare records: 26


,date,temperature,humidity,pressure,wind_speed,precipitation,station
0,2025-01-01 00:00:00,6.854305,22.514335,1012.101582,1.292043,1.031239,KORD
12,2025-01-01 12:00:00,27.459919,94.375812,1012.598063,0.633769,9.506071,KORD
15,2025-01-01 15:00:00,-1.746797,89.716847,991.862454,17.399355,4.484455,KORD
16,2025-01-01 16:00:00,3.690901,84.293766,996.269985,10.223824,2.932108,KORD
22,2025-01-01 22:00:00,3.146509,45.440278,984.683738,16.504934,0.912061,KORD


In [7]:
# Multiple stations
major_airports = df.query("station in ['KJFK', 'KORD']")

print(f"Major airports records: {len(major_airports)}")
print("\nStation distribution:")
print(major_airports['station'].value_counts())

Major airports records: 48

Station distribution:
station
KORD    26
KJFK    22
Name: count, dtype: int64


### 6. Precipitation Events
Identify different precipitation intensities:

In [8]:
# Light rain (0. 1-2. 5 mm/hr)
light_rain = df.query('0.1 <= precipitation < 2.5')

# Heavy rain (>7. 6 mm/hr)
heavy_rain = df.query('precipitation >= 7.6')

# No precipitation
dry_conditions = df.query('precipitation == 0')

print(f"Light rain events: {len(light_rain)}")
print(f"Heavy rain events: {len(heavy_rain)}")
print(f"Dry conditions: {len(dry_conditions)}")

Light rain events: 25
Heavy rain events: 29
Dry conditions: 0


### 7. High Pressure Systems
Filter for anticyclonic conditions:

In [9]:
# High pressure with stable conditions
high_pressure = df.query('pressure > 1013 and wind_speed < 8')

print(f"High pressure systems: {len(high_pressure)} records")
print(f"\nAverage conditions during high pressure:")
print(high_pressure[['temperature', 'humidity', 'pressure', 'wind_speed']]. mean())

High pressure systems: 12 records

Average conditions during high pressure:
temperature       9.472613
humidity         50.448214
pressure       1021.667004
wind_speed        2.919171
dtype: float64


### 8. Storm Conditions
Identify potential severe weather:

In [10]:
# Low pressure with high winds
storm_potential = df.query('pressure < 990 and wind_speed > 15')

print(f"Storm potential records: {len(storm_potential)}")
storm_potential.head()

Storm potential records: 8


,date,temperature,humidity,pressure,wind_speed,precipitation,station
5,2025-01-01 05:00:00,-2.980247,39.943378,980.459853,24.396302,8.955232,KMSP
8,2025-01-01 08:00:00,17.050176,38.303853,980.253079,19.879655,9.053820,KSEA
22,2025-01-01 22:00:00,3.146509,45.440278,984.683738,16.504934,0.912061,KORD
37,2025-01-02 13:00:00,-5.604755,76.241517,981.215798,17.591446,8.218601,KJFK
83,2025-01-04 11:00:00,-7.139874,90.187148,984.699097,21.936800,6.275581,KSEA


## Advanced Examples

### 9. Using Variables
Pass Python variables into queries using `@`:

In [11]:
# Define thresholds
temp_threshold = 20
wind_threshold = 12

# Use variables in query
filtered = df.query('temperature > @temp_threshold and wind_speed < @wind_threshold')

print(f"Filtered records:  {len(filtered)}")
print(f"Criteria:  temperature > {temp_threshold}°C and wind_speed < {wind_threshold} m/s")
filtered.head()

Filtered records:  20
Criteria:  temperature > 20°C and wind_speed < 12 m/s


,date,temperature,humidity,pressure,wind_speed,precipitation,station
7,2025-01-01 07:00:00,28.977927,80.444091,1013.175088,8.073912,0.108377,KMSP
9,2025-01-01 09:00:00,21.863266,26.158393,988.040403,6.770806,0.912867,KSEA
11,2025-01-01 11:00:00,33.645943,32.897703,1014.594760,1.961410,9.500620,KMSP
12,2025-01-01 12:00:00,27.459919,94.375812,1012.598063,0.633769,9.506071,KORD
33,2025-01-02 09:00:00,32.699849,47.009214,1004.625885,2.901816,1.011227,KORD


### 10. Heat Index Calculations
Combine conditions to identify uncomfortable conditions:

In [12]:
# Hot and humid (feels-like temperature warning)
uncomfortable = df.query('temperature > 27 and humidity > 70')

print(f"Uncomfortable conditions: {len(uncomfortable)} records")
uncomfortable.head()

Uncomfortable conditions: 6 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station
1,2025-01-01 01:00:00,32.782144,70.912833,984.206998,13.283866,9.025529,KJFK
7,2025-01-01 07:00:00,28.977927,80.444091,1013.175088,8.073912,0.108377,KMSP
12,2025-01-01 12:00:00,27.459919,94.375812,1012.598063,0.633769,9.506071,KORD
34,2025-01-02 10:00:00,33.453441,95.432776,989.762149,1.150066,0.841068,KMSP
50,2025-01-03 02:00:00,33.631308,92.661271,994.722445,12.578406,7.771469,KORD


### 11. Inversions and Stable Conditions
Identify temperature inversion potential:

In [13]:
# Cold, calm conditions favorable for inversions
inversion_conditions = df.query('temperature < 5 and wind_speed < 3 and pressure > 1020')

print(f"Inversion potential:  {len(inversion_conditions)} records")
inversion_conditions.head()

Inversion potential:  3 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station
56,2025-01-03 08:00:00,-6.017837,73.770844,1026.807739,0.662783,0.113536,KSEA
71,2025-01-03 23:00:00,-1.057794,21.327026,1020.505670,0.455546,9.698788,KORD
72,2025-01-04 00:00:00,-9.751505,60.967445,1023.353616,2.361074,0.431599,KORD


### 12. Negation (NOT)
Exclude certain conditions:

In [14]:
# Everything except Seattle station
not_seattle = df.query("station != 'KSEA'")

print(f"Non-Seattle records: {len(not_seattle)}")
print("\nStation distribution:")
print(not_seattle['station'].value_counts())

Non-Seattle records: 86

Station distribution:
station
KMSP    38
KORD    26
KJFK    22
Name: count, dtype: int64


In [15]:
# Not in a list
coastal_stations = df.query("station not in ['KSEA', 'KJFK']")

print(f"Non-coastal station records: {len(coastal_stations)}")
print("\nStation distribution: ")
print(coastal_stations['station'].value_counts())

Non-coastal station records: 64

Station distribution: 
station
KMSP    38
KORD    26
Name: count, dtype: int64


### 13. Complex Boolean Logic
Combine multiple conditions with parentheses:

In [16]:
# Either foggy conditions OR storm conditions
hazardous = df.query('(humidity > 95 and wind_speed < 2) or (wind_speed > 20 and precipitation > 5)')

print(f"Hazardous conditions: {len(hazardous)} records")
hazardous.head()

Hazardous conditions: 13 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station
5,2025-01-01 05:00:00,-2.980247,39.943378,980.459853,24.396302,8.955232,KMSP
13,2025-01-01 13:00:00,-0.444740,84.649630,991.213465,24.066210,5.734379,KJFK
14,2025-01-01 14:00:00,-1.817876,70.672301,1015.608961,20.899503,6.318372,KMSP
34,2025-01-02 10:00:00,33.453441,95.432776,989.762149,1.150066,0.841068,KMSP
51,2025-01-03 03:00:00,24.880977,39.164951,999.254886,21.412246,5.584042,KMSP


### 14. Date/Time Filtering
Filter by specific date criteria:

In [17]:
# First, ensure date is datetime type
df['date'] = pd.to_datetime(df['date'])

# Extract hour and day components
df['hour'] = df['date'].dt.hour
df['day'] = df['date'].dt.day

# Filter by specific date criteria
morning_data = df.query('hour >= 6 and hour <= 10')

print(f"Morning data (6-10 AM): {len(morning_data)} records")
morning_data.head()

Morning data (6-10 AM): 20 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station,hour,day
6,2025-01-01 06:00:00,-7.386237,52.830634,985.073577,12.907509,3.892017,KSEA,6,1
7,2025-01-01 07:00:00,28.977927,80.444091,1013.175088,8.073912,0.108377,KMSP,7,1
8,2025-01-01 08:00:00,17.050176,38.303853,980.253079,19.879655,9.053820,KSEA,8,1
9,2025-01-01 09:00:00,21.863266,26.158393,988.040403,6.770806,0.912867,KSEA,9,1
10,2025-01-01 10:00:00,-9.073698,43.180116,1007.436689,10.974286,3.193136,KMSP,10,1


In [18]:
# Mid-month data
mid_month = df.query('day >= 10 and day <= 20')

print(f"Mid-month data:  {len(mid_month)} records")
mid_month.head()

Mid-month data:  0 records


,date,temperature,humidity,pressure,wind_speed,precipitation,station,hour,day


## Performance Tips

### 15. Column Name Spaces
If your column names have spaces, use backticks to specify the column name:

In [19]:
# Create a copy with column name containing spaces
df_with_spaces = df.copy()
df_with_spaces['dew point'] = np.random.uniform(-5, 20, len(df))

# Query using backticks for column names with spaces
result = df_with_spaces.query('`dew point` > 15')

print(f"Records with dew point > 15°C: {len(result)}")
result[['temperature', 'humidity', 'dew point']].head()

Records with dew point > 15°C: 13


,temperature,humidity,dew point
1,32.782144,70.912833,16.304538
4,-2.979161,92.605318,16.916340
18,9.437526,91.404720,17.603967
24,10.523149,38.234813,16.559093
25,25.332918,54.168623,18.738016


### 16. Method Chaining
Combine query with other Pandas operations:

In [22]:
# Filter and then calculate statistics
result = (df.query('temperature > 20 and station == "KORD"')
            .groupby('station')['humidity']
            .agg(['mean', 'min', 'max', 'count']))

print("Statistics for warm days at Chicago O'Hare:")
result

Statistics for warm days at Chicago O'Hare:


,mean,min,max,count
station,,,,
KORD,65.209794,45.856235,94.375812,7


## Common Meteorological Use Cases

### 17. Freezing Level Identification

In [ ]:
# Near-freezing conditions (important for precipitation type)
freezing_level = df.query('-2 <= temperature <= 2')

print(f"Near-freezing conditions: {len(freezing_level)} records")
freezing_level.head()

### 18. Wind Chill Conditions

In [ ]:
# Cold and windy (wind chill warning criteria)
wind_chill_warning = df.query('temperature < -10 and wind_speed > 10')

print(f"Wind chill warning conditions: {len(wind_chill_warning)} records")
wind_chill_warning.head()

### 19. Drought Monitoring

In [ ]:
# Extended dry periods with low humidity
drought_conditions = df.query('precipitation < 0.1 and humidity < 40')

print(f"Drought conditions: {len(drought_conditions)} records")
drought_conditions.head()

### 20. Perfect Storm Setup

In [ ]:
# Conditions favorable for severe weather development
severe_setup = df.query('pressure < 1000 and temperature > 25 and humidity > 60 and wind_speed > 12')

print(f"Severe weather setup: {len(severe_setup)} records")
severe_setup.head()

## Comparison:  query() vs Boolean Indexing

### Traditional Boolean Indexing

In [ ]:
# Traditional way
result_traditional = df[(df['temperature'] > 20) & (df['humidity'] < 60)]

print(f"Records using traditional indexing: {len(result_traditional)}")
result_traditional.head()

### Using query() (cleaner and more readable)

In [ ]:
# Using query()
result_query = df.query('temperature > 20 and humidity < 60')

print(f"Records using query(): {len(result_query)}")
result_query.head()

In [ ]:
# Verify both methods give the same result
print(f"Both methods return same results: {result_traditional.equals(result_query)}")

## Key Advantages of query()

1. **Readability**: More natural syntax, especially for complex conditions
2. **Performance**: Can be faster for large DataFrames
3. **Cleaner code**: Less parentheses and ampersands
4. **String-based**: Easy to construct queries dynamically

## Common Pitfalls

1. **Column names with special characters**: Use backticks
2. **String values**: Must use quotes inside the query string
3. **Boolean operators**: Use `and`, `or`, `not` (not `&`, `|`, `~`)
4. **Variables**:  Remember to use `@` prefix

# Practice Exercises

Try these on your own in the cells below:

### Exercise 1
Find all records where temperature is between 10-20°C and wind speed exceeds 15 m/s

In [ ]:
# Your code here


### Exercise 2
Identify conditions where humidity is below 30% (dry conditions)

In [ ]:
# Your code here


### Exercise 3
Filter for high pressure (>1020 hPa) at KMSP station only

In [ ]:
# Your code here


### Exercise 4
Find potential thunderstorm conditions:  temp > 25°C, humidity > 70%, and pressure dropping below 1010 hPa

In [ ]:
# Your code here


## Summary Statistics

Let's look at some overall statistics from our dataset:

In [20]:
# Overall dataset statistics
print("Dataset Summary Statistics:")
print("="*60)
df[['temperature', 'humidity', 'pressure', 'wind_speed', 'precipitation']].describe()

Dataset Summary Statistics:


,temperature,humidity,pressure,wind_speed,precipitation
count,100.000000,100.000000,100.000000,100.000000,100.000000
mean,11.158133,59.826538,1005.880067,12.278724,5.160458
std,13.387023,23.448900,14.671312,7.336305,3.186009
min,-9.751505,20.556170,980.253079,0.359837,0.108377
25%,-1.305966,39.360362,993.843993,6.240373,2.639348
50%,10.886410,60.449989,1008.127747,12.742958,5.253995
75%,22.859140,81.294688,1017.618347,18.394438,7.968842
max,34.409912,98.852036,1029.502693,24.762629,9.929648


In [21]:
# Station distribution
print("\nObservations by Station:")
print("="*60)
df['station'].value_counts()


Observations by Station:


station
KMSP    38
KORD    26
KJFK    22
KSEA    14
Name: count, dtype: int64

## Resources

- [Pandas query() documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html)
- Remember: `query()` is great for filtering, but for complex transformations, use other Pandas methods

---

Happy data filtering! 🌤️📊